core/server.go

```diff
@@ -437,6 +440,29 @@ func (m *RpcServer) DestroyEnvironment(cxt context.Context, req *pb.DestroyEnvir
 	return m.doTeardownAndCleanup(env, req.Force, req.KeepTasks)
 }
 
+func (m *RpcServer) SetEnvironmentName(cxt context.Context, req *pb.SetEnvironmentNameRequest) (resp *pb.Empty, err error) {
+	m.logMethod()
+
+	if req == nil || len(req.Id) == 0 {
+		return nil, status.New(codes.InvalidArgument, "received nil request").Err()
+	}
+
+	var envId uid.ID
+	envId, err = uid.FromString(req.Id)
+	if err != nil {
+		return nil, status.New(codes.InvalidArgument, "received bad environment id").Err()
+	}
+
+	var env *environment.Environment
+	env, err = m.state.environments.Environment(envId)
+	if err != nil {
+		return nil, status.Newf(codes.NotFound, "environment not found: %s", err.Error()).Err()
+	}
+
+	env.SetName(req.GetName())
```

Comment:
The first 3 failure modes here are correct and I 100% agree with them, but we are missing a 4th one: if the user tries to set an environment name that already identifies another environment. We must do this in order to guarantee the uniqueness of env names, and since a name is not a UUID, we don't get the guarantee of uniqueness for free, and we must implement it ourselves.

I propose that the `env.SetName` function is made to return an error, and then you can forward this error by returning a gRPC `status` here. Inside `SetName`, you'd then iterate over all envs to check their names before going ahead with the write.

Depending on how you implement this, you might need to either jump one level up (for example a `environment.manager.SetEnvName(envId, newName)`) or pass a handle to the envMan to the existing function (e.g. `env.SetName(envMan, req.GetName())`). Let me know if you'd like us to discuss this further.